From efc9bdc0c1f94e30cd33332fc5ed2f5749293aab Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Sat, 15 Jun 2024 02:51:36 +0200
Subject: [PATCH 16/23] Add Scrypt hash function (#2)

Scrypt is the PoW algorithm used by dogecoin.

This has simply been copied from the dogecoin source with some refurbishments.
---
 src/crypto/CMakeLists.txt |   1 +
 src/crypto/scrypt.cpp     | 285 ++++++++++++++++++++++++++++++++++++++
 src/crypto/scrypt.h       |  45 ++++++
 src/test/CMakeLists.txt   |   1 +
 src/test/scrypt_tests.cpp |  75 ++++++++++
 5 files changed, 407 insertions(+)
 create mode 100644 src/crypto/scrypt.cpp
 create mode 100644 src/crypto/scrypt.h
 create mode 100644 src/test/scrypt_tests.cpp

diff --git a/src/crypto/CMakeLists.txt b/src/crypto/CMakeLists.txt
index cb377582485a5f6654e3b6d1a9cfe5e85f8d7591..741f825b6f04206f45e53b3a40683f8bb4cc3be4 100644
--- a/src/crypto/CMakeLists.txt
+++ b/src/crypto/CMakeLists.txt
@@ -14,6 +14,7 @@ add_library(crypto
 	muhash.cpp
 	poly1305.cpp
 	ripemd160.cpp
+	scrypt.cpp
 	sha1.cpp
 	sha256.cpp
 	sha256_sse4.cpp
diff --git a/src/crypto/scrypt.cpp b/src/crypto/scrypt.cpp
new file mode 100644
index 0000000000000000000000000000000000000000..f56269f64b1fb61c08150cbd763b21880d856090
--- /dev/null
+++ b/src/crypto/scrypt.cpp
@@ -0,0 +1,285 @@
+/*
+ * Copyright 2009 Colin Percival, 2011 ArtForz, 2012-2013 pooler
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ * This file was originally written by Colin Percival as part of the Tarsnap
+ * online backup system.
+ */
+
+#include <crypto/hmac_sha256.h>
+#include <crypto/scrypt.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#if defined(USE_SSE2) && !defined(USE_SSE2_ALWAYS)
+#ifdef _MSC_VER
+// MSVC 64bit is unable to use inline asm
+#include <intrin.h>
+#else
+// GCC Linux or i686-w64-mingw32
+#include <cpuid.h>
+#endif
+#endif
+
+#ifndef __FreeBSD__
+
+static inline void be32enc(void *pp, uint32_t x) {
+    uint8_t *p = (uint8_t *)pp;
+    p[3] = x & 0xff;
+    p[2] = (x >> 8) & 0xff;
+    p[1] = (x >> 16) & 0xff;
+    p[0] = (x >> 24) & 0xff;
+}
+
+#endif
+
+static inline uint32_t le32dec(const void *pp) {
+    const uint8_t *p = (uint8_t const *)pp;
+    return ((uint32_t)(p[0]) + ((uint32_t)(p[1]) << 8) +
+            ((uint32_t)(p[2]) << 16) + ((uint32_t)(p[3]) << 24));
+}
+
+static inline void le32enc(void *pp, uint32_t x) {
+    uint8_t *p = (uint8_t *)pp;
+    p[0] = x & 0xff;
+    p[1] = (x >> 8) & 0xff;
+    p[2] = (x >> 16) & 0xff;
+    p[3] = (x >> 24) & 0xff;
+}
+
+/**
+ * PBKDF2_SHA256(passwd, passwdlen, salt, saltlen, c, buf, dkLen):
+ * Compute PBKDF2(passwd, salt, c, dkLen) using HMAC-SHA256 as the PRF, and
+ * write the output to buf.  The value dkLen must be at most 32 * (2^32 - 1).
+ */
+void PBKDF2_SHA256(const uint8_t *passwd, size_t passwdlen, const uint8_t *salt,
+                   size_t saltlen, uint64_t c, uint8_t *buf, size_t dkLen) {
+    CHMAC_SHA256 baseCtx(passwd, passwdlen);
+    CHMAC_SHA256 hctx(nullptr, 0);
+    size_t i;
+    uint8_t ivec[4];
+    uint8_t U[CHMAC_SHA256::OUTPUT_SIZE];
+    uint8_t T[CHMAC_SHA256::OUTPUT_SIZE];
+    uint64_t j;
+    unsigned int k;
+    size_t clen;
+
+    /* Compute HMAC state after processing P and S. */
+    CHMAC_SHA256 PShctx = baseCtx;
+    PShctx.Write(salt, saltlen);
+
+    /* Iterate through the blocks. */
+    for (i = 0; i * 32 < dkLen; i++) {
+        /* Generate INT(i + 1). */
+        be32enc(ivec, (uint32_t)(i + 1));
+
+        /* Compute U_1 = PRF(P, S || INT(i)). */
+        hctx = PShctx;
+        hctx.Write(ivec, 4);
+        hctx.Finalize(U);
+
+        /* T_i = U_1 ... */
+        memcpy(T, U, CHMAC_SHA256::OUTPUT_SIZE);
+
+        for (j = 2; j <= c; j++) {
+            /* Compute U_j. */
+            hctx = baseCtx;
+            hctx.Write(U, CHMAC_SHA256::OUTPUT_SIZE);
+            hctx.Finalize(U);
+
+            /* ... xor U_j ... */
+            for (k = 0; k < CHMAC_SHA256::OUTPUT_SIZE; k++) {
+                T[k] ^= U[k];
+            }
+        }
+
+        /* Copy as many bytes as necessary into buf. */
+        clen = dkLen - i * CHMAC_SHA256::OUTPUT_SIZE;
+        if (clen > CHMAC_SHA256::OUTPUT_SIZE) {
+            clen = CHMAC_SHA256::OUTPUT_SIZE;
+        }
+        memcpy(&buf[i * CHMAC_SHA256::OUTPUT_SIZE], T, clen);
+    }
+}
+
+#define ROTL(a, b) (((a) << (b)) | ((a) >> (32 - (b))))
+
+static inline void xor_salsa8(uint32_t B[16], const uint32_t Bx[16]) {
+    uint32_t x00, x01, x02, x03, x04, x05, x06, x07, x08, x09, x10, x11, x12,
+        x13, x14, x15;
+    int i;
+
+    x00 = (B[0] ^= Bx[0]);
+    x01 = (B[1] ^= Bx[1]);
+    x02 = (B[2] ^= Bx[2]);
+    x03 = (B[3] ^= Bx[3]);
+    x04 = (B[4] ^= Bx[4]);
+    x05 = (B[5] ^= Bx[5]);
+    x06 = (B[6] ^= Bx[6]);
+    x07 = (B[7] ^= Bx[7]);
+    x08 = (B[8] ^= Bx[8]);
+    x09 = (B[9] ^= Bx[9]);
+    x10 = (B[10] ^= Bx[10]);
+    x11 = (B[11] ^= Bx[11]);
+    x12 = (B[12] ^= Bx[12]);
+    x13 = (B[13] ^= Bx[13]);
+    x14 = (B[14] ^= Bx[14]);
+    x15 = (B[15] ^= Bx[15]);
+    for (i = 0; i < 8; i += 2) {
+        /* Operate on columns. */
+        x04 ^= ROTL(x00 + x12, 7);
+        x09 ^= ROTL(x05 + x01, 7);
+        x14 ^= ROTL(x10 + x06, 7);
+        x03 ^= ROTL(x15 + x11, 7);
+
+        x08 ^= ROTL(x04 + x00, 9);
+        x13 ^= ROTL(x09 + x05, 9);
+        x02 ^= ROTL(x14 + x10, 9);
+        x07 ^= ROTL(x03 + x15, 9);
+
+        x12 ^= ROTL(x08 + x04, 13);
+        x01 ^= ROTL(x13 + x09, 13);
+        x06 ^= ROTL(x02 + x14, 13);
+        x11 ^= ROTL(x07 + x03, 13);
+
+        x00 ^= ROTL(x12 + x08, 18);
+        x05 ^= ROTL(x01 + x13, 18);
+        x10 ^= ROTL(x06 + x02, 18);
+        x15 ^= ROTL(x11 + x07, 18);
+
+        /* Operate on rows. */
+        x01 ^= ROTL(x00 + x03, 7);
+        x06 ^= ROTL(x05 + x04, 7);
+        x11 ^= ROTL(x10 + x09, 7);
+        x12 ^= ROTL(x15 + x14, 7);
+
+        x02 ^= ROTL(x01 + x00, 9);
+        x07 ^= ROTL(x06 + x05, 9);
+        x08 ^= ROTL(x11 + x10, 9);
+        x13 ^= ROTL(x12 + x15, 9);
+
+        x03 ^= ROTL(x02 + x01, 13);
+        x04 ^= ROTL(x07 + x06, 13);
+        x09 ^= ROTL(x08 + x11, 13);
+        x14 ^= ROTL(x13 + x12, 13);
+
+        x00 ^= ROTL(x03 + x02, 18);
+        x05 ^= ROTL(x04 + x07, 18);
+        x10 ^= ROTL(x09 + x08, 18);
+        x15 ^= ROTL(x14 + x13, 18);
+    }
+    B[0] += x00;
+    B[1] += x01;
+    B[2] += x02;
+    B[3] += x03;
+    B[4] += x04;
+    B[5] += x05;
+    B[6] += x06;
+    B[7] += x07;
+    B[8] += x08;
+    B[9] += x09;
+    B[10] += x10;
+    B[11] += x11;
+    B[12] += x12;
+    B[13] += x13;
+    B[14] += x14;
+    B[15] += x15;
+}
+
+void scrypt_1024_1_1_256_sp_generic(const uint8_t *input, uint8_t *output,
+                                    uint8_t *scratchpad) {
+    uint8_t B[128];
+    uint32_t X[32];
+    uint32_t *V;
+    uint32_t i, j, k;
+
+    V = (uint32_t *)(((uintptr_t)(scratchpad) + 63) & ~(uintptr_t)(63));
+
+    PBKDF2_SHA256(input, 80, input, 80, 1, B, 128);
+
+    for (k = 0; k < 32; k++) {
+        X[k] = le32dec(&B[4 * k]);
+    }
+
+    for (i = 0; i < 1024; i++) {
+        memcpy(&V[i * 32], X, 128);
+        xor_salsa8(&X[0], &X[16]);
+        xor_salsa8(&X[16], &X[0]);
+    }
+    for (i = 0; i < 1024; i++) {
+        j = 32 * (X[16] & 1023);
+        for (k = 0; k < 32; k++)
+            X[k] ^= V[j + k];
+        xor_salsa8(&X[0], &X[16]);
+        xor_salsa8(&X[16], &X[0]);
+    }
+
+    for (k = 0; k < 32; k++) {
+        le32enc(&B[4 * k], X[k]);
+    }
+
+    PBKDF2_SHA256(input, 80, B, 128, 1, output, 32);
+}
+
+#if defined(USE_SSE2)
+// By default, set to generic scrypt function. This will prevent crash in case
+// when scrypt_detect_sse2() wasn't called
+void (*scrypt_1024_1_1_256_sp_detected)(const uint8_t *input, uint8_t *output,
+                                        uint8_t *scratchpad) =
+    &scrypt_1024_1_1_256_sp_generic;
+
+void scrypt_detect_sse2() {
+#if defined(USE_SSE2_ALWAYS)
+    printf("scrypt: using scrypt-sse2 as built.\n");
+#else // USE_SSE2_ALWAYS
+    // 32bit x86 Linux or Windows, detect cpuid features
+    unsigned int cpuid_edx = 0;
+#if defined(_MSC_VER)
+    // MSVC
+    int x86cpuid[4];
+    __cpuid(x86cpuid, 1);
+    cpuid_edx = (unsigned int)buffer[3];
+#else  // _MSC_VER
+    // Linux or i686-w64-mingw32 (gcc-4.6.3)
+    unsigned int eax, ebx, ecx;
+    __get_cpuid(1, &eax, &ebx, &ecx, &cpuid_edx);
+#endif // _MSC_VER
+
+    if (cpuid_edx & 1 << 26) {
+        scrypt_1024_1_1_256_sp_detected = &scrypt_1024_1_1_256_sp_sse2;
+        printf("scrypt: using scrypt-sse2 as detected.\n");
+    } else {
+        scrypt_1024_1_1_256_sp_detected = &scrypt_1024_1_1_256_sp_generic;
+        printf("scrypt: using scrypt-generic, SSE2 unavailable.\n");
+    }
+#endif // USE_SSE2_ALWAYS
+}
+#endif
+
+void scrypt_1024_1_1_256(const uint8_t *input, uint8_t *output) {
+    thread_local uint8_t scratchpad[SCRYPT_SCRATCHPAD_SIZE];
+    memset(scratchpad, 0, sizeof(scratchpad));
+    scrypt_1024_1_1_256_sp(input, output, scratchpad);
+}
diff --git a/src/crypto/scrypt.h b/src/crypto/scrypt.h
new file mode 100644
index 0000000000000000000000000000000000000000..3df2304e503370928d83b7c1ba493de634c93a9d
--- /dev/null
+++ b/src/crypto/scrypt.h
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2009 Colin Percival, 2011 ArtForz, 2012-2013 pooler
+ * All rights reserved.
+ */
+
+#ifndef BITCOIN_CRYPTO_SCRYPT_H
+#define BITCOIN_CRYPTO_SCRYPT_H
+
+#include <stdint.h>
+#include <stdlib.h>
+
+static const int SCRYPT_SCRATCHPAD_SIZE = 131072 + 63;
+
+void scrypt_1024_1_1_256(const uint8_t *input, uint8_t *output);
+void scrypt_1024_1_1_256_sp_generic(const uint8_t *input, uint8_t *output,
+                                    uint8_t *scratchpad);
+
+#if defined(USE_SSE2)
+
+#if defined(_M_X64) || defined(__x86_64__) || defined(_M_AMD64) ||             \
+    (defined(MAC_OSX) && defined(__i386__))
+#define USE_SSE2_ALWAYS 1
+#define scrypt_1024_1_1_256_sp(input, output, scratchpad)                      \
+    scrypt_1024_1_1_256_sp_sse2((input), (output), (scratchpad))
+#else
+#define scrypt_1024_1_1_256_sp(input, output, scratchpad)                      \
+    scrypt_1024_1_1_256_sp_detected((input), (output), (scratchpad))
+#endif
+
+void scrypt_detect_sse2();
+void scrypt_1024_1_1_256_sp_sse2(const uint8_t *input, uint8_t *output,
+                                 uint8_t *scratchpad);
+extern void (*scrypt_1024_1_1_256_sp_detected)(const uint8_t *input,
+                                               uint8_t *output,
+                                               uint8_t *scratchpad);
+#else
+#define scrypt_1024_1_1_256_sp(input, output, scratchpad)                      \
+    scrypt_1024_1_1_256_sp_generic((input), (output), (scratchpad))
+
+#endif // defined(USE_SSE2)
+
+void PBKDF2_SHA256(const uint8_t *passwd, size_t passwdlen, const uint8_t *salt,
+                   size_t saltlen, uint64_t c, uint8_t *buf, size_t dkLen);
+
+#endif // BITCOIN_CRYPTO_SCRYPT_H
diff --git a/src/test/CMakeLists.txt b/src/test/CMakeLists.txt
index 9c5a6db6a0e66ba26d43e7a8ed696b924ebbd87e..093e683d9367db7f3f5d4831e9b807eb0b050fa2 100644
--- a/src/test/CMakeLists.txt
+++ b/src/test/CMakeLists.txt
@@ -230,6 +230,7 @@ add_boost_unit_tests_to_suite(bitcoin test_bitcoin
 		scriptnum_tests.cpp
 		scriptnum_63bit_tests.cpp
 		scriptinterpreter_tests.cpp
+		scrypt_tests.cpp
 		serialize_tests.cpp
 		serialize_intcode_tests.cpp
 		settings_tests.cpp
diff --git a/src/test/scrypt_tests.cpp b/src/test/scrypt_tests.cpp
new file mode 100644
index 0000000000000000000000000000000000000000..6a4ab347f3c4cc1e8a882fe7377d1526fe045798
--- /dev/null
+++ b/src/test/scrypt_tests.cpp
@@ -0,0 +1,75 @@
+// Copyright (c) 2024 The Bitcoin developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <boost/test/unit_test.hpp>
+
+#include <crypto/scrypt.h>
+#include <uint256.h>
+#include <util/strencodings.h>
+
+BOOST_AUTO_TEST_SUITE(scrypt_tests)
+
+BOOST_AUTO_TEST_CASE(scrypt_hashtest) {
+    // Test Scrypt hash with known inputs against expected outputs
+
+    struct TestCase {
+        std::string input;
+        std::string output;
+    };
+
+    const std::vector<TestCase> TEST_CASES = {
+        {
+            "020000004c1271c211717198227392b029a64a7971931d351b387bb80db027f270"
+            "411e398a07046f7d4a08dd815412a8712f874a7ebf0507e3878bd24e20a3b73fd7"
+            "50a667d2f451eac7471b00de6659",
+            "00000000002bef4107f882f6115e0b01f348d21195dacd3582aa2dabd7985806",
+        },
+        {
+            "0200000011503ee6a855e900c00cfdd98f5f55fffeaee9b6bf55bea9b852d9de2c"
+            "e35828e204eef76acfd36949ae56d1fbe81c1ac9c0209e6331ad56414f9072506a"
+            "77f8c6faf551eac7471b00389d01",
+            "00000000003a0d11bdd5eb634e08b7feddcfbbf228ed35d250daf19f1c88fc94",
+        },
+        {
+            "02000000a72c8a177f523946f42f22c3e86b8023221b4105e8007e59e81f6beb01"
+            "3e29aaf635295cb9ac966213fb56e046dc71df5b3f7f67ceaeab24038e743f883a"
+            "ff1aaafaf551eac7471b0166249b",
+            "00000000000b40f895f288e13244728a6c2d9d59d8aff29c65f8dd5114a8ca81",
+        },
+        {
+            "010000007824bc3a8a1b4628485eee3024abd8626721f7f870f8ad4d2f33a27155"
+            "167f6a4009d1285049603888fe85a84b6c803a53305a8d497965a5e896e1a00568"
+            "359589faf551eac7471b0065434e",
+            "00000000003007005891cd4923031e99d8e8d72f6e8e7edc6a86181897e105fe",
+        },
+        {
+            "0200000050bfd4e4a307a8cb6ef4aef69abc5c0f2d579648bd80d7733e1ccc3fbc"
+            "90ed664a7f74006cb11bde87785f229ecd366c2d4e44432832580e0608c579e4cb"
+            "76f383f7f551eac7471b00c36982",
+            "000000000018f0b426a4afc7130ccb47fa02af730d345b4fe7c7724d3800ec8c",
+        }};
+
+#if defined(USE_SSE2)
+    scrypt_detect_sse2();
+#endif
+
+    uint8_t scratchpad[SCRYPT_SCRATCHPAD_SIZE];
+    for (const TestCase &test_case : TEST_CASES) {
+        uint256 scrypthash;
+        std::vector<uint8_t> input = ParseHex(test_case.input);
+
+#if defined(USE_SSE2)
+        // Test SSE2 scrypt
+        scrypt_1024_1_1_256_sp_sse2(&input[0], scrypthash.data(), scratchpad);
+        BOOST_CHECK_EQUAL(scrypthash.ToString().c_str(), test_case.output);
+#endif
+
+        // Test generic scrypt
+        scrypt_1024_1_1_256_sp_generic(&input[0], scrypthash.data(),
+                                       scratchpad);
+        BOOST_CHECK_EQUAL(scrypthash.ToString().c_str(), test_case.output);
+    }
+}
+
+BOOST_AUTO_TEST_SUITE_END()
